### PR TITLE
feat(hax-lib): intro. `hax` feature

### LINKED ISSUE #639: Hax-lib[-macros]: `cfg(hax)`, `cfg(debug_assert)` and release mode policy
hax-lib and hax-lib-macros provide hax-related features that are relevant only when extracting though hax (and maybe in some scenarios in debug mode, but this is not entirely clear yet)

The constraint here is production code. When we introduce a dependency to hax-lib in production code, we want (if not `cfg(hax)`):
 - as few as possible dependencies (from hax-lib);
 - all macros from `hax-lib` should be no-ops.

Solution: add a dummy `hax-lib-dummy` crate that is an empty shell, providing the same interface to `hax-lib`, but with macros being no-ops. Then `hax-lib`, with a `cfg`, either includes the real library or the dummy one.

Actions:
 - [ ] Rename `hax-lib` into `hax-lib-impl`, move it inside a brand new `hax-lib` crate
 - [ ] Create a `hax-lib-dummy` crate
 - [ ] Make `hax-lib` expose either the dummy one or the lib one
 - [ ] Move all docstrings to external files so that the dummy and impl crate have the same documentation

### PR DESCRIPTION
**Pull request by W95Psp**
This PR introduces a (default) `hax` feature on `hax-lib`.

When this feature is disabled:
 - `hax-lib-macros` defines identity proc-macros, and keeps its dependencies to a minimum (only `quote` and `syn`)
 - `hax-lib` becomes a one-module tiny and "inert" library, with only `hax-lib-macros` as a dependency (when the feature `macros` is enabled)

Fixes #639.

### CODE REVIEW

**Review by karthikbhargavan — APPROVED** [review]
I tested this on libcrux and it appears to work well.

**Comment by karthikbhargavan** [comment]
@franziskuskiefer  fixed the flags on this PR, and now the hax engine fails because of this error:

```
# File "lib/concrete_ident_generated.ml", lines 1-476, characters 0-94:
#   1 | type t = 
#   2 |       Hax_lib_protocol__crypto__Impl_5__from_bytes
#   3 |     | Core__result__Impl__E
#   4 |     | Core__ops__control_flow__ControlFlow__Break
#   5 |     | Core__ops__bit__BitAnd__bitand
# ...
# 473 |     | Core__cmp__PartialEq__eq__'__1
# 474 |     | Rust_primitives__hax__control_flow_monad__moption
# 475 |     | Core__array__iter__Impl__T
# 476 |     | Core__iter__traits__iterator__Iterator[@@deriving show, yojson, compare, sexp, eq, hash]
# Error: Two constructors are named Hax_lib__assert
```

This is because the hax-engine also depends on hax-lib to extract names, and with the new cfg flags, something is going wrong in the way this is used.

@W95Psp: maybe this will be easy for you to fix.

**Review by franziskuskiefer — COMMENTED** [review]
Clearing the review. Re-request when it's ready to look at.

File: hax-lib/macros/src/dummy.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    quote! { () }.into()
+}
+
+#[proc_macro_attribute]
+pub fn lemma(_attr: TokenStream, _item: TokenStream) -> TokenStream {
+    quote! {}.into()
+}
+
+fn unsafe_expr() -> TokenStream {
+    quote! { ::std::compile_error!("`hax_lib::unsafe_expr` has no meaning outside of hax extraction, please use it solely on hax only places.") }.into()
```
[the remark this answers is not in the document]
Comment (W95Psp) [reply hax-lib/macros/src/dummy.rs]: To make the public interfaces of the dummy and real libraries the same (ideally the rustdoc output of both should be the same)

[the remark this answers is not in the document]
Comment (W95Psp) [reply hax-lib/macros/src/dummy.rs]: Added an explanation in 40d14c09b11da16d71742f48371ef4d5f8fc69c1

**Comment by W95Psp** [comment]
The engine is fixed and the library is fixed as well, things are passing CI.
@karthikbhargavan can you check if that's working fine on MLKEM?

**Review by franziskuskiefer — APPROVED** [review]
Looks good to me now. The main thing I checked is that we are really not getting any dependencies or code when not using the hax cfg.

**Comment by karthikbhargavan** [comment]
libcrux proofs use `hax_lib::loop_invariant!`; so this needs to be in the dummy-lib.

**Comment by W95Psp** [comment]
I fixed it, now it should be fine!

**Review by karthikbhargavan — APPROVED** [review]
This works for me now. Thanks!

**Pull request merged**